### PR TITLE
openldap: adding PIC for lining against openldap static libs from a s…

### PR DESCRIPTION
**Pull request by thierryba**
As it is linux specific, the home-brew guys asked to submit the patch here.
Essentially this allows the static libs to linked into a shared libs (I was building apache's mod_ldap).

### CODE REVIEW

File: Formula/openldap.rb [hunk 1]
```diff
@@ -45,6 +45,7 @@ def install
       --enable-translucent
       --enable-unique
       --enable-valsort
+      --enable-pic
```
Comment (iMichka) [inline Formula/openldap.rb]: We prefer to keep Linux related stuff separated form upstream's code, because it makes us the life easier when we merge homebrew-core into linuxbrew-core.

That would be something like:
`args << "--enable-pic" unless OS.mac?`

Could you change the code accordingly?

**Comment by iMichka** [comment]
Hi. Welcome on Linuxbrew. I made a small comment, nothing big.

Do new bottles need to be rebuilt? In that case you will need to add the following line below the sha256 line: `revision 1 unless OS.mac?`

**Comment by thierryba** [comment]
@iMichka: Thank you. I made the change you suggested.

**Comment by sjackman** [comment]
Hi, Thierry. Thanks for your first contribution to Linuxbrew. Could you please give an example of your original use case that failed without this PR? A `brew gist-logs` showing the error is best if possible.

**Comment by thierryba** [comment]
Hi Shaun,

For me it was failing outside of linuxbrew. But I was trying to build apache against static openldap libs and when it comes to mod_ldap it fails because the code is not relocatable.

Regards,
Thierry

**Comment by sjackman** [comment]
Does it work if you instead build against the dynamic openldap libs rather than the static ones. Typically static libraries are non-PIC, and dynamic libraries are PIC.

**Comment by thierryba** [comment]
Or does indeed but post of the live out there are pic. For instance openssl
is. And in this case they have a build option for that. Maybe it could be a
brew option?

Le 2 sept. 2017 2:23 AM, "Shaun Jackman" <notifications@github.com> a
écrit :

**Comment by sjackman** [comment]
> Or does indeed but post of the live out there are pic.

Sorry, Thierry. I didn't understand. Could you rephrase it?

> Maybe it could be a brew option?

Options aren't tested by CI, so we try to avoid them unless there's a strong use case for them.

**Comment by thierryba** [comment]
Sorry I was on my phone and French keyboard (with French autocorrect….) not a good mix.
So what I was saying is that lots of libs (for instance OpenSSL) enable PIC by default. Here I’m trying to simplify my build process by relying on home/linuxbrew to create an apache server that is relatively easily usable in a package. So linking against openldap shared libs is too much of a headache for me.
So I thought we could add that option in. It does not hurt anyone.
The other option would be to have that as a build option. But as you mentioned, options is probably overkill for that little feature.

**Comment by sjackman** [comment]
We prefer to avoid diverging from upstream Homebrew/core unless there's a strong use case. We merge Homebrew/core into Linuxbrew/core daily, and resolving merge conflicts takes developer time. You can if you like maintain a private tap with a few modified formulae (like `openldap`), and pin that tap so that it takes precedence over the core repo.

**Comment by iMichka** [comment]
Same opinion. We would like to not diverge too much from upstream. I do not really get the point why this is a Linux specific problem? If you want that option (be it an option or a default build flag), you need to lobby upstream to add it.

Of course the solution to have your own tap with the modified formula is also nice, in case you do not get that flag merged upstream. I'll close here without merging, but feel free to comment on this issue.

**Comment by sjackman** [comment]
Here's an example of a separate tap that provides a bunch of statically-linked executables: https://github.com/Homebrew/homebrew-portable/tree/master/Formula

**Comment by sjackman** [comment]
I stumbled on this argument explaining the cost of -fPIC by chance today:

> Do not use -fPIC globally! While I understand that this works around an ancient CentOS bug, enabling -fPIC for static libraries on x86 causes undue register pressure, due to the need for keeping the Global Offset Table (GOT) address in a register. Instead, CMake knows that only for shared objects, -fPIC has to be added, but not for static libraries. See https://nebelwelt.net/publications/files/12TRpie.pdf for the toll position independent code takes on x86. Again, if for security reasons of address space layout randomization (ASLR), users can always add -fPIC -fPIE to CXXFLAGS is they so desire.

https://github.com/pezmaster31/bamtools/pull/139
